c10/core/TensorImpl.cpp

```diff
@@ -401,7 +415,9 @@ void TensorImpl::copy_tensor_metadata(
     const c10::VariableVersion& version_counter,
     bool allow_tensor_metadata_change) {
   copy_tensor_metadata_except_version_counter(src_impl, dest_impl, allow_tensor_metadata_change);
-  dest_impl->set_version_counter(version_counter);
+  if (!dest_impl->is_inference_tensor()) {
+    dest_impl->set_version_counter(version_counter);
```

Comment:
Perhaps this is another one for subsequent refactoring, but this also looks a bit suspicious, because if I'm copying the VC from source to dest, then if source was an inference tensor the VC is empty and it's harmless to set version counter here. This should be a consequence of https://github.com/pytorch/pytorch/pull/54403/files#r607472380

I think one reason this might not hold is because of this indirect call of copy tensor metadata

```
Tensor VariableHooks::variable_data(const Tensor& self) const {
  TORCH_CHECK(self.defined(), "cannot call variable_data() on undefined tensor");
  auto self_impl_copy = self.unsafeGetTensorImpl()->shallow_copy_and_detach(
    /*version_counter=*/0,
    /*allow_tensor_metadata_change=*/false);
```

so this is freshly creating a new version counter and would trigger the error here. This is kind of suspicious anyway because we don't actually want to allocate a version counter and then throw it out, so maybe just testing if self is an inference tensor in this function would solve this problem.

Reply:
Yea this one will be considered in the subsequent refactor, essentially we have some code calling `shallow_copy_and_detach` and they're all suspicious. E.g. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.h#L688 is called when you jit load a model inside InferenceMode and cause here to fail. 
In `as_view` we also call `shallow_copy_and_detach` so the followup refactor will focus on rationalize uses `shallow_copy_and_detach` and the VariableVersions passed from there.